File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    h = len(triangle)
    dp = [[0 for _ in range(h+1)] for i in range(h+1)]
    
    for i, l in enumerate(triangle):
        for j, num in enumerate(l):
            dp[i+1][j+1] = max(dp[i][j], dp[i][j+1]) + triangle[i][j]
    return max(dp[h])

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.*;
class Solution {
    public int solution(int[][] triangle) {
        int size = triangle.length;
    int[][] dp = new int[2][size+1];
    dp[0][1] = triangle[0][0];

    for (int turn = 0; turn < size - 1; turn++) {
      for (int x = 1; x <= triangle[turn+1].length; x++) {
        dp[1][x] = Math.max(dp[0][x-1], dp[0][x]) + triangle[turn+1][x-1];
      }
      for (int x = 1; x <= size; x++) {
        dp[0][x] = dp[1][x];
      }
    }
    int max = dp[1][1];
    for (int i = 1; i <= size; i++) {
      if (max < dp[1][i]) max = dp[1][i];
    }
    return max;
    }
}